Add /contract endpoints

chris124567 opened the pull request:
This adds a `muse` like API to the new siad.  This PR depends on code implemented in https://github.com/SiaFoundation/core/pull/24 so we should wait to merge until that is finalized.

lukechampine commented on api/api.go:
```diff
@@ -46,3 +47,37 @@ type SyncerPeer struct {
+type ContractBaseRequest struct {
+	NetAddress string          `json:"netAddress"`
+	TheirKey   types.PublicKey `json:"theirKey"`
+}
```
I'd just call this `ScanRequest`, and inline the fields in `FormRequest` and `RenewRequest`

lukechampine commented on api/api.go:
```diff
@@ -46,3 +47,37 @@ type SyncerPeer struct {
+type ContractRenewRequest struct {
+	ContractBaseRequest
+	RenterKey             types.PrivateKey            `json:"renterKey"`
+	Contract              *types.FileContractRevision `json:"contract"`
```
why use a pointer here?

lukechampine commented on api/client.go:
```diff
@@ -104,6 +105,24 @@ func (c *Client) SyncerConnect(addr string) (err error) {
 }
 
+// ContractScan scans a host and retrieves its settings.
+func (c *Client) ContractScan(netAddress string, theirKey types.PublicKey) (resp rhp.HostSettings, err error) {
```
`theirKey` should be `hostKey` (here and elsewhere)

lukechampine commented on api/client.go:
```diff
@@ -104,6 +105,24 @@ func (c *Client) SyncerConnect(addr string) (err error) {
+
+// ContractForm forms a contract.
+func (c *Client) ContractForm(netAddress string, theirKey types.PublicKey, renterKey types.PrivateKey, hostFunds, renterFunds types.Currency, endHeight uint64) (resp ContractFormResponse, err error) {
+	err = c.post("/contract/form", ContractFormRequest{ContractBaseRequest{netAddress, theirKey}, renterKey, hostFunds, renterFunds, endHeight}, &resp)
```
this is a pretty large struct; best to write it out one-line-per-field

chris124567 commented:
Rebased and added support for host sets.  I just left the import to go.sia.tech/core/renter even though it is no longer present because a similar thing (after it was decided that specific RPC handling code like that shouldn't go in `core`) is being worked on in siad and I will replace those references once it is ported.

chris124567 commented:
Rebased to support the recent renterd PR.  Moved all the API methods to the renterd api module.  The level of code duplication between renterd/siad makes me uncomfortable but I guess there may not be a better way.

lukechampine commented on api/renterd/api.go:
```diff
@@ -49,6 +57,10 @@ type server struct {
+
+	mu        sync.Mutex
+	contracts []rhp.Contract
+	hostSets  map[string][]types.PublicKey
```
for now, we can just return the contract in the API response and not save anything server-side. We can skip the host set stuff too; I'm not sure it's the best design

lukechampine commented on api/renterd/api.go:
```diff
@@ -188,6 +200,144 @@ func (s *server) syncerConnectHandler(w http.ResponseWriter, req *http.Request,
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
```
we'll want to add a `defer session.Close()` here

chris124567 replied:
As far as I can tell Session does not have a Close method

lukechampine commented on api/renterd/api.go:
```diff
@@ -210,5 +360,13 @@ func NewServer(cm ChainManager, s Syncer, w Wallet, tp TransactionPool) http.Han
 
+	mux.POST("/contract/scan", srv.contractScanHandler)
+	mux.POST("/contract/form", srv.contractFormHandler)
+	mux.POST("/contract/renew", srv.contractRenewHandler)
```
needs an `/api` prefix

lukechampine commented on api/renterd/requests.go:
```diff
@@ -44,3 +47,49 @@ type SyncerPeerResponse struct {
+type ContractFormResponse struct {
+	Contract rhp.Contract        `json:"contract"`
+	Parents  []types.Transaction `json:"parents"`
+}
```
My gut says that renew should have a separate `Response` type, even if they currently contain the same fields.

chris124567 commented:
Tried to resolve the merge conflicts with the Github website and it did this... one second.

lukechampine commented:
It's actually fine (and preferable) to define `ChainManager` in both places; the real `*chain.Manager` will satisfy both interfaces.

chris124567 commented:
These errors are strange.  I don't get them on my system (go1.16.10 linux/amd64) and it doesn't really make sense to me why 

```
assignment mismatch: 1 variable but s.cm.TipContext returns 2 values
```

is the error for this line (which assigns two variables):

```
    vc, err := s.cm.TipContext()
```
when all definitions of ChainManager are:

```
	// A ChainManager manages blockchain state.
	ChainManager interface {
		TipContext() (consensus.ValidationContext, error)
	}
```

chris124567 commented:
Finally :) Should be good to go.
